{
    "before_functions": "public UnixUser authenticate(String username, String password) throws PAMException {\n    this.password = password;\n    try {\n        check(libpam.pam_set_item(pht,PAM_USER,username),\"pam_set_item failed\");\n        check(libpam.pam_authenticate(pht,0),\"pam_authenticate failed\");\n        check(libpam.pam_setcred(pht,0),\"pam_setcred failed\");\n        // several different error code seem to be used to represent authentication failures\n//            check(libpam.pam_acct_mgmt(pht,0),\"pam_acct_mgmt failed\");\n\n        PointerByReference r = new PointerByReference();\n        check(libpam.pam_get_item(pht,PAM_USER,r),\"pam_get_item failed\");\n        String userName = r.getValue().getString(0);\n        passwd pwd = libc.getpwnam(userName);\n        if(pwd==null)\n            throw new PAMException(\"Authentication succeeded but no user information is available\");\n        return new UnixUser(userName,pwd);\n    } finally {\n        this.password = null;\n    }\n}\n",
    "after_functions": "public UnixUser authenticate(String username, String password) throws PAMException {\n    this.password = password;\n    try {\n        check(libpam.pam_set_item(pht,PAM_USER,username),\"pam_set_item failed\");\n        check(libpam.pam_authenticate(pht,0),\"pam_authenticate failed\");\n        check(libpam.pam_setcred(pht,0),\"pam_setcred failed\");\n        check(libpam.pam_acct_mgmt(pht,0),\"pam_acct_mgmt failed\");\n\n        PointerByReference r = new PointerByReference();\n        check(libpam.pam_get_item(pht,PAM_USER,r),\"pam_get_item failed\");\n        String userName = r.getValue().getString(0);\n        passwd pwd = libc.getpwnam(userName);\n        if(pwd==null)\n            throw new PAMException(\"Authentication succeeded but no user information is available\");\n        return new UnixUser(userName,pwd);\n    } finally {\n        this.password = null;\n    }\n}\n"
}